data_model/derive/src/lib.rs

```diff
@@ -51,38 +44,50 @@ impl syn::parse::Parse for Items {
 /// }
 ///
 /// /* will produce:
-/// pub struct DataModel1 {
-///    #[cfg(feature = "transparent_api")]
-///    pub item1: u32,
-///    #[cfg(not(feature = "transparent_api"))]
-///    pub(crate) item1: u32,
-///    item2: u64
-/// }
+/// pub mod model {
+///     pub struct DataModel1 {
+///         #[cfg(feature = "transparent_api")]
+///         pub item1: u32,
+///         #[cfg(not(feature = "transparent_api"))]
+///         pub(crate) item1: u32,
+///         pub(super) item2: u64
+///     }
 ///
-/// #[cfg(not(feature = "transparent_api"))]
-/// pub struct DataModel2 {
-///    pub item1: u32,
-///    item2: u64
-/// }
+///     #[cfg(not(feature = "transparent_api"))]
+///     pub struct DataModel2 {
+///         pub item1: u32,
+///         pub(super) item2: u64
+///     }
 ///
-/// #[cfg(feature = "transparent_api")]
-/// struct DataModel2 {
-///    pub item1: u32,
-///    item2: u64
+///     #[cfg(feature = "transparent_api")]
+///     struct DataModel2 {
+///         pub item1: u32,
+///         pub(super) item2: u64
+///     }
 /// }
 /// */
 /// ```
-#[proc_macro]
+#[proc_macro_attribute]
 #[proc_macro_error::proc_macro_error]
-pub fn model(input: TokenStream) -> TokenStream {
-    let input = parse_macro_input!(input as Items);
-    let mut items = Vec::new();
+pub fn model(_attr: TokenStream, input: TokenStream) -> TokenStream {
+    model::impl_model(&parse_macro_input!(input)).into()
+}
 
-    for item in input.0 {
-        items.push(api::process_item(item));
-    }
+/// Same as [`model`] macro, but only processes a single item.
+///
+/// You should prefer using [`model`] macro over this one.
+#[proc_macro]
+#[proc_macro_error::proc_macro_error]
+pub fn model_single(input: TokenStream) -> TokenStream {
+    model::process_item(parse_macro_input!(input)).into()
+}
 
-    quote::quote! { #(#items)* }.into()
+/// Stub for usage inside [`model`] macro.
+/// Will make [`model`] macro to produce the same item but with
+/// [`iroha_ffi_derive::FfiType`] when `ffi_export` feature is enabled and `ffi_import` is disabled.
+#[proc_macro_attribute]
+pub fn ffi_type(_attr: TokenStream, _input: TokenStream) -> TokenStream {
```

Comment:
why do we have this?

Reply:
Because without this it will be enable to compile code inside `model` mod which uses `#[ffi_type]`. I just replicated the same behaviour as before. This annotation will be processed inside `#[model]` macro

Reply:
> This annotation will be processed inside #[model] macro

doesn't this processing remove this attribute? I don't think it has to be declared as a proc macro

Reply:
No, looks like `rustc` wants `#[ffi_type]` to be declared before `#[model]` expansion

Reply:
does this mean that macros inside the module are expanded first? and module level macros are expanded afterwards

Reply:
I don't think so. I think it's like derive attributes. `rustc` wont allow to use undefined attributes despite that any of derive macros can use it

Reply:
interesting. And we can't move the `ffi_type` code into this function because then we would have to make sure it's top-level macro? I would panic inside this function then because it can only be used when wrapped in `model!` macro

Reply:
Hm, good idea, I'll try this

Reply:
It worked!

Reply:
I was wrong. I was meslead by another error which was fixed by that. Now everything works without it